subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/codegen/GradleApiExtensionsIntegrationTest.kt

```diff
@@ -0,0 +1,202 @@
+                    }
+                }
+            }
+        """)
+
+        withBuildScript("""
+            container(String::class)
+            apply(from = "plugin.gradle.kts")
+        """)
+
+        withFile("plugin.gradle.kts", """
+            import org.apache.tools.ant.filters.ReplaceTokens
+
+            // Class<T> to KClass<T>
+            property(Long::class)
+
+            // Groovy named arguments to vararg of Pair
+            fileTree("dir" to "src", "excludes" to listOf("**/ignore/**", "**/.data/**"))
+
+            // Class<T> + Action<T> to KClass<T> + T.() -> Unit
+            tasks.register("foo", Copy::class, Action<Copy> {
+                from("src")
+                into("dst")
+
+                // Class<T> + Groovy named arguments to KClass<T> + vararg of Pair
+                filter(ReplaceTokens::class, "foo" to "bar")
+            })
+        """)
+
+        build("foo", "-I", "init.gradle.kts")
+    }
+
+    @Test
+    fun `can use Gradle API generated extensions in buildSrc`() {
+
+        withSettingsIn("buildSrc", """
+            pluginManagement {
+                repositories {
+                    gradlePluginPortal()
+                    kotlinDev()
+                }
+            }
+        """)
+        withBuildScriptIn("buildSrc", """
+            plugins {
+                `kotlin-dsl`
+                `java-gradle-plugin`
+            }
+            apply<org.gradle.kotlin.dsl.plugins.precompiled.PrecompiledScriptPlugins>()
+        """)
+
+        withFile("buildSrc/src/main/kotlin/foo/FooTask.kt", """
+            package foo
+
+            import org.gradle.api.*
+            import org.gradle.api.tasks.*
+
+            import org.gradle.kotlin.dsl.*
+
+            import org.apache.tools.ant.filters.ReplaceTokens
+
+            open class FooTask : DefaultTask() {
+
+                @TaskAction
+                fun foo() = project.run {
+                    container(Long::class)
+                }
+            }
+        """)
+
+        withFile("buildSrc/src/main/kotlin/foo/foo.gradle.kts", """
+            package foo
+
+            tasks.register("foo", FooTask::class)
+        """)
+
+        withBuildScript("""
+            plugins {
+                id("foo.foo")
+            }
+        """)
+
+        build("foo")
+    }
+
+    @Test
+    fun `generated jar contains Gradle API extensions sources and byte code`() {
+
+        val tag = "gradleUserHomeDir="
+
+        withBuildScript("""
+            task("printGradleUserHomeDir") {
+                doLast {
+                    println("$tag${'$'}{gradle.gradleUserHomeDir}")
+                }
+            }
+        """)
+
+        val gradleUserHomeDir = build("printGradleUserHomeDir", "-q").output.lines()
+            .single { it.contains(tag) }.substringAfter(tag).let { File(it) }
+
+        val generatedJar = gradleUserHomeDir.resolve("caches")
+            .listFiles { f -> f.isDirectory && f.name[0].isDigit() }.single()
```

Comment:
⭕️ This looks very fragile and in fact, after a few iterations the `single()` call started failing.

Reply:
Should be much more solid now